### PR TITLE
TASK-3: OOP Principles

### PR DESCRIPTION
**Pull request by DimaLayeuski**
NO-TASK: Added gitignore

### CODE REVIEW

**Comment by DaryaNikitina** [comment]
Remove obj, bin and .idea directories. Add sln file.

**Comment by DimaLayeuski** [comment]
> Remove obj, bin and .idea directories. Add sln file.

I have a problem with file from solution. I try to fix that today.

File: AQA_Lab_Task/Program.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+﻿using Driver;
+
+var driverGenerator = new DriverGenerator();
+var drivers = driverGenerator.GenerateDrivers();
+var vehicleGenerator = new VehicleGenerator(drivers);
+var vehicles = vehicleGenerator.GenerateVehicles();
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Program.cs]: If you need to generate drivers and vehicles count as 2,5,10, you will go inside DriverGenerator and VehicleGenerator - modify count variable.
Create a variable and pass count through constructor here or better to pass in Generate methods.

File: AQA_Lab_Task/Program.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+foreach (var driver in drivers)
+{
+    Console.WriteLine($"Driver {drivers.IndexOf(driver) + 1}");
+    driver.Print();
+}
+
+
+Console.Write("\nChoose driver (1-3): ");
+var choice = Console.ReadLine();
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Program.cs]: Make a check for incorrect input. If its incorrect - ask user to choose driver again

File: AQA_Lab_Task/Program.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        var drivingExperience = DateTime.Today.Year - chosenDriver.DateDriverLicense.Year;
+        Console.WriteLine($"Driving experience is {drivingExperience}");
+        foreach (var vehicle in ownedVehicles)
+        {
+            {
+                Console.Write($"Enter mileage of {vehicle.Model}: ");
+                var mileage = int.Parse(Console.ReadLine() ?? string.Empty);
+                var fuelConsumption = vehicle.Engine.AverageConsumption * mileage / 100;
+                var time = mileage / vehicle.Engine.MaximumSpeed;
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Program.cs]: time is 0 every run

File: AQA_Lab_Task/Program.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                Console.WriteLine($"Fuel consumption for {vehicle.Model}: {fuelConsumption:F} l");
+                Console.WriteLine($"Drive time with max speed: {time} h\n");
+            }
+        }
+
+        break;
+    }
+}
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Program.cs]: We have only one attempt to see information. Make it in loop, where we can back and choose another option for information.

File: AQA_Lab_Task/Vehicle.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+namespace Driver;
+
+public class Vehicle
+{
+    public string? Model { get; set; }
+    private int Year { get; set; }
+    public Driver? Owner { get; set; }
+    public Engine Engine { get; set; }
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Vehicle.cs]: Make all fields private.

File: AQA_Lab_Task/Truck.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+namespace Driver;
+
+public class Truck : Vehicle
+{
+    public bool HasPricep { get; set; }
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Truck.cs]: Make a field private.

File: AQA_Lab_Task/SportCar.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+namespace Driver;
+
+public class SportCar : Vehicle
+{
+    public SportCar(int year, string model, Engine engine, Driver? person) : base(year, model, engine)
+    {
+        Owner = HasExperience(person) ? person : null;
+    }
+
+    private static bool HasExperience(Driver? person)
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/SportCar.cs]: For private method, which you are calling inside the class - dont use static.

File: AQA_Lab_Task/Person.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+namespace Driver;
+
+public class Person
+{
+    public string? FirstName { get; set; } = default!;
+    public string? LastName { get; set; } = default!;
+    public DateTime DateOfBirth { get; set; } = default!;
+    public bool IsDriver { get; set; } = default!;
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Person.cs]: Make all fields private. Remove = default!

File: AQA_Lab_Task/Minivan.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+namespace Driver;
+
+public class Minivan : Vehicle
+{
+    public int SeatsCount;
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Minivan.cs]: Make a field private.

File: AQA_Lab_Task/Engine.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
+namespace Driver;
+
+public class Engine
+{
+    public double Capacity { get; set; } = default!;
+    public int Power { get; set; } = default!;
+    public FuelTypes FuelType { get; set; } = default!;
+    public int MaximumSpeed { get; set; } = default!;
+    public double AverageConsumption { get; set; } = default!;
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Engine.cs]: Make all fields private. Remove = default!

File: AQA_Lab_Task/DriverGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+                .RuleFor(x => x.DateOfBirth, x => x.Person.DateOfBirth.Date)
+                .RuleFor(x => x.IdNumber, _ => Guid.NewGuid())
+                .RuleFor(x => x.IsDriver, x => x.Random.Bool())
+                .RuleFor(x => x.DateDriverLicense,
+                    x => x.Date.Between(x.Person.DateOfBirth.Date.AddYears(16), DateTime.Now))
+                .Generate();
+
+            _drivers.Add(driver);
+        }
+
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/DriverGenerator.cs]: Remove spaces(Empty lines)

File: AQA_Lab_Task/Driver.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+namespace Driver;
+
+public class Driver : Person
+{
+    public DateTime DateDriverLicense { get; set; } = default!;
+    public Guid IdNumber { get; set; } = default!;
+    public Vehicle Vehicle { get; set; } = null!;
```
Comment (DaryaNikitina) [inline AQA_Lab_Task/Driver.cs]: Make all fields private. Remove = default!
Remove unused field

**Review by DaryaNikitina — CHANGES_REQUESTED** [review]

**Pull request closed without merging**